### PR TITLE
Build Linux, Mac, and Windows binaries

### LINKED ISSUE #929: Provide binaries for acs-engine
**Is this a request for help?**:
No
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):

Feature Request

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
Any

**What happened**:
I want to install acs-engine locally to try to use it (since `az acs` seems to lag behind significantly). However, I can't find binaries that I can download and install. `go get` works but uses master, which I am wary of using.

**What you expected to happen**:
I go to the 'releases' tab and find binaries I can download of acs-engine for the latest version.

**How to reproduce it** (as minimally and precisely as possible):
Go to releases tab. Although there are releases, there isn't any binaries

### PR DESCRIPTION
**Pull request by seanknox**
**What this PR does / why we need it**: This PR provides the ability to build acs-engine binaries for the following platforms/architectures:
- darwin/amd64 
- linux/amd64 
- windows/amd64 

Additionally it adds make tasks to create distribution archives for each platform/architecture. Also improved the the bootstrap step (now called `make bootstrap`) to be more intelligent about installing development dependencies.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #929 

**Special notes for your reviewer**:

**Release note**:

```release-note 
add `make build-cross dist` to build binaries and archives
```

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/1077)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -13,15 +23,40 @@ GOFILES=`go list ./... | grep -v "github.com/Azure/acs-engine/vendor" | sed 's|g
...
+# usage: make clean build-cross dist VERSION=v0.4.0
+.PHONY: build-cross
+build-cross: LDFLAGS += -extldflags "-static"
+build-cross:
+	CGO_ENABLED=0 gox -output="_dist/{{.OS}}-{{.Arch}}/{{.Dir}}" -osarch='$(TARGETS)' $(GOFLAGS) -tags '$(TAGS)' -ldflags '$(LDFLAGS)'
+
+.PHONY: dist
+dist:
+	( \
+		cd _dist && \
```
Comment (slack) [inline Makefile]: After a `make clean` the `_dist` dir needs to be re-created.

Comment (slack) [inline Makefile]: Ignore me, was not running the full chain! Looks like `build-cross` creates _dist as a side effect from gox.

**Comment by slack** [comment]
LGTM, I ran the commands and everything looks good.

File: Makefile [hunk 2]
```diff
@@ -1,9 +1,19 @@
+TARGETS           = darwin/amd64 linux/amd64 windows/amd64
```
Comment (JiangtianLi) [inline Makefile]: For windows binary, is it build inside docker container or Windows OS?

Comment (JiangtianLi) [inline Makefile]: Looks you are using cross-compile on Linux/docker container for Windows binary. LGTM.

Comment (seanknox) [inline Makefile]: I'm using [gox](https://github.com/mitchellh/gox) as a cross-build wrapper for go. The `build-cross` task doesn't use Docker, but a CI step definitely could. Do you mind giving an 👍 ?

Comment (JiangtianLi) [inline Makefile]: awesome

Comment (JiangtianLi) [inline Makefile]: awesome

**Comment by seanknox** [comment]
Hey @JiangtianLi, can you approve this PR if you are happy with it?

**Comment by karataliu** [comment]
Is following line in Makefile in use?
```
PKG       := $(shell glide novendor)
```

When using `make bootstrapp` to install glide, it will result in 'make: glide: Command not found' first.

**Comment by seanknox** [comment]
That's copy/paste cruft that should be removed. Nothing refers to `PKG`. Thanks for spotting.

**Pull request merged**